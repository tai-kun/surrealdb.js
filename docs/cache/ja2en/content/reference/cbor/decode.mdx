---
title: decode
---

## Synchronous Decoding

The `decode` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decode(input: Uint8Array, options?: DecodeOptions): unknown;
```

#### Arguments

##### `input`

The CBOR-formatted `Uint8Array` to decode into a JavaScript value.

##### `options`

Decoding options.

`mapType?: "Object" | "Map"`
:   This option can be used to choose whether to decode key-value objects as JavaScript plain objects or `Map` objects. Choose `"Object"` to decode as a plain object and `"Map"` to decode as a `Map` class instance. The default is `"Object"`. Plain object keys can only be strings or numbers (i.e., `string | number`), but if you decode with this option set to `"Map"`, the keys can be any decoded value.

`reviver?: Reviver | ReviverObject | (Reviver | ReviverObject)[]`
:   This option is used to transform CBOR simple data items or tagged data items into JavaScript values. Simple data items are limited to values from 0 to 19 and 32 to 255. The tag of a tagged data item is a `number` or `bigint` from 0 to 2^64-1. If the function returns the `CONTINUE` symbol, which can be obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions will be called. If the value cannot be transformed, `SurrealTypeError` is thrown. 

`isSafeMapKey?: (key: unknown, map: Map<unknown, unknown>) => boolean`
:   This option is a function that validates whether a key of a `Map` object is safe. The function takes two arguments: `key` and `map`. `key` is the value of the key, and `map` is the `Map` object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: Record<string | number, unknown>) => boolean`
:   This option is a function that validates whether a key of a plain object is safe. The function takes two arguments: `key` and `obj`. `key` is the value of the key, and `obj` is the plain object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of the object. The depth is incremented by 1 for each object or array entered. The default value is `64`. Since recursion is not used for decoding, there is no upper limit to the depth, but it is recommended to keep it small.

#### Return Value

The JavaScript value decoded from the CBOR-formatted `Uint8Array`.

### Example

#### Decoding as a `Map` object

```ts
import { decode } from "@tai-kun/surrealdb/cbor";

const cbor = new Uint8Array([
  0xa1, // mt: 5, ai: 1
  // key
  0x00, // mt: 0, ai: 0
  // value
  0x00, // mt: 0, ai: 0
]);

const value = decode(bytes, { mapType: "Map" });

console.log(value); // Map(1) { [ 0 ] => 0 }
```

## Asynchronous Decoding

The `decodeStream` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decodeStream(
  input: ReadableStream<Uint8Array>,
  options?: DecodeStreamOptions,
): unknown;
```

#### Arguments

##### `input`

A readable stream of `Uint8Array`. The Fetch API may provide this stream on the `body` property of the `Response` object in response to a request.

##### `options`

In addition to the options for synchronous decoding, the following options can be specified:

`signal?: AbortSignal`
:   An AbortSignal to abort the stream.

#### Return Value

A `StatefulPromise` that resolves with the JavaScript value.

### Example

#### Fetch API

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";

const response = await fetch("http://localhost:8000/rpc", {
  method: "POST",
  headers: {
    Accept: "application/cbor",
    // ... more headers
  },
  // ... more properties
})

const result = await decodeStream(resp.body);
```
