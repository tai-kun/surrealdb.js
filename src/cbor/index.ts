export type * from "./decode-stream";
export { default as decodeStream } from "./decode-stream";

export type * from "./decode";
export { default as decode } from "./decode";

export {
  CONTINUE,
  type IsSafeMapKey,
  type IsSafeObjectKey,
  type Reviver,
  type ReviverObject,
  type SimpleItemReviver,
  type TaggedItemReviver,
} from "./decoder";

export type * from "./encode";
export { default as encode } from "./encode";

export {
  type AdditionalInfo,
  AI_EIGHT_BYTES,
  AI_FOUR_BYTES,
  AI_INDEFINITE_NUM_BYTES,
  AI_ONE_BYTE,
  AI_SIMPLE_FALSE,
  AI_SIMPLE_NULL,
  AI_SIMPLE_TRUE,
  AI_SIMPLE_UNDEFINED,
  AI_TWO_BYTES,
  type DataItem,
  type DataItemValue,
  HEADER_FALSE,
  HEADER_FLOAT_DOUBLE,
  HEADER_FLOAT_HALF,
  HEADER_NULL,
  HEADER_TRUE,
  HEADER_UNDEFINED,
  type MajorType,
  MT_ARRAY,
  MT_BYTE_STRING,
  MT_MAP,
  MT_NEGATIVE_INTEGER,
  MT_SIMPLE_FLOAT,
  MT_TAG,
  MT_UNSIGNED_INTEGER,
  MT_UTF8_STRING,
  Simple,
} from "./spec";

export type * from "./tagged";
export { default as Tagged } from "./tagged";

export type { ToCBOR } from "./traits";
export { canToCBOR } from "./traits";

export {
  type Replacer,
  write,
  writeBigInt,
  writeBoolean,
  writeByteString,
  writeEncodedUtf8String,
  writeHeader,
  writeNullable,
  writeNumber,
  writePayload,
  writeUtf8String,
} from "./write-utils";

export type * from "./writer"; // `Writer` も型だけでよし
