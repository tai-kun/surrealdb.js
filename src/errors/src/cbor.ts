import { SurrealError, type SurrealErrorOptions } from "./general";

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborerror)
 */
export class CborError extends SurrealError {
  static {
    this.prototype.name = "CborError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborwellformednesserror)
 */
export class CborWellFormednessError extends CborError {
  static {
    this.prototype.name = "CborWellFormednessError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoomuchdataerror)
 */
export class CborTooMuchDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooMuchDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super("Unconsumed input bytes remain after decoding.", options);
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoolittledataerror)
 */
export class CborTooLittleDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooLittleDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super(
      "Input data appears truncated or incomplete for CBOR decoding.",
      options,
    );
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborsyntaxerror)
 */
export class CborSyntaxError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborSyntaxError";
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbormaxdepthreachederror)
 */
export class CborMaxDepthReachedError extends CborError {
  static {
    this.prototype.name = "CborMaxDepthReachedError";
  }

  constructor(
    public maxDepth: number,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(`Maximum depth of ${maxDepth} has been reached.`, options);
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborunsafemapkeyerror)
 */
export class CborUnsafeMapKeyError extends CborError {
  static {
    this.prototype.name = "CborUnsafeMapKeyError";
  }

  constructor(
    public key: unknown,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(
      `Invalid key for CBOR map: ${String(key)}. The key must be a valid CBOR`
        + " data type that is safe to use in a JavaScript map or object.",
      options,
    );
  }
}

export class CborDecodeStreamAbortFailedError extends CborError {
  static {
    this.prototype.name = "CborStreamAbortFailedError";
  }

  override cause: unknown[];

  constructor(
    errors: readonly unknown[],
    options?: Omit<SurrealErrorOptions, "cause"> | undefined,
  ) {
    super("Failed to abort decode-stream", options);
    this.cause = errors.slice();
  }
}

export class CborMemoryError extends CborError {
  static {
    this.prototype.name = "CborMemoryError";
  }
}

export class CborMemoryBlockError extends CborError {
  static {
    this.prototype.name = "CborMemoryBlockError";
  }
}

type MemoryAddress = string | number;

export class CborMemoryBlockConflictError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockConflictError";
  }

  constructor(
    address: MemoryAddress,
    expectedSize: number,
    actualSize: number,
  ) {
    super(
      `Memory conflict at address ${address}: `
        + `Expected size ${expectedSize}, but found ${actualSize}.`,
    );
  }
}

export class CborMemoryBlockInUseError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockInUseError";
  }
}

export class CborUndefinedMemoryBlockError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborUndefinedMemoryBlockError";
  }

  constructor(address: MemoryAddress) {
    super(`Memory block at address ${address} is not defined.`);
  }
}
