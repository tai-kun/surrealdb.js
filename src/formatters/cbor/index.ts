export type * from "./formatter";
export { default } from "./formatter";
