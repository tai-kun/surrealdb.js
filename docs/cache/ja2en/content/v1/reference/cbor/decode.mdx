---
title: Decode
slug: en/v1/reference/cbor/decode
---

## Synchronous Decoding

The `decode` function decodes a `Uint8Array` in CBOR format into a JavaScript value.

### Import

```ts
import { decode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decode(input: Uint8Array, options?: DecodeOptions): unknown;
```

#### Arguments

##### `input`

A `Uint8Array` in CBOR format to be decoded into a JavaScript value.

##### `options`

Decoding options.

`mapType?: "Object" | "Map"`
:   This option can be used to choose whether to decode key-value objects as JavaScript plain objects or `Map` objects. Selecting `"Object"` decodes as a plain object, while `"Map"` decodes as an instance of the `Map` class. The default value is `"Object"`. Plain object keys can only accept strings or numbers (i.e. `string | number`), but decoding with this option set to `"Map"` allows keys to accept all decoded values.

`reviver?: Reviver | ReviverObject | (Reviver | ReviverObject)[]`
:   This option is used to convert simple or tagged data items from CBOR to JavaScript values. Simple data items are limited to values between 0 and 19 and between 32 and 255. The tag of a tagged data item is a `number` or `bigint` between 0 and 2^64-1. If the function returns the `CONTINUE` symbol obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions are called. If no conversion could be made, a `SurrealTypeError` is thrown.

`isSafeMapKey?: (key: unknown, map: Map<unknown, unknown>) => boolean`
:   This option is a function that verifies whether a key in a `Map` object is safe. The function accepts two arguments: `key`, which is the key value, and `map`, which is the `Map` object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` is thrown. By default, it determines `"__proto__"` or `"constructor"` as unsafe keys. Please note that specifying this option completely overrides the default validation. If you want to add your own validation to the default validation, you should also determine `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: Record<string | number, unknown>) => boolean`
:   This option is a function that verifies whether a key in a plain object is safe. The function accepts two arguments: `key`, which is the key value, and `obj`, which is the plain object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` is thrown. By default, it determines `"__proto__"` or `"constructor"` as unsafe keys. Please note that specifying this option completely overrides the default validation. If you want to add your own validation to the default validation, you should also determine `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of objects. The depth increases by 1 for every object or array entry. The default value is `64`. While decoding does not use recursive processing, there is no upper limit on depth, but it is recommended that it not be too large.

#### Return value

The JavaScript value decoded from the `Uint8Array` in CBOR format.

### Example

#### Decoding with `Map` objects

```ts
import { decode } from "@tai-kun/surrealdb/cbor";

const cbor = new Uint8Array([
  0xa1, // mt: 5, ai: 1
  // key
  0x00, // mt: 0, ai: 0
  // value
  0x00, // mt: 0, ai: 0
]);

const value = decode(bytes, { mapType: "Map" });

console.log(value); // Map(1) { [ 0 ] => 0 }
```

## Asynchronous Decoding

The `decodeStream` function decodes a `Uint8Array` in CBOR format into a JavaScript value.

### Import

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decodeStream(
  input: ReadableStream<Uint8Array>,
  options?: DecodeStreamOptions,
): unknown;
```

#### Arguments

##### `input`

A readable stream of `Uint8Array`. The Fetch API may provide this stream in the `body` property of the `Response` object for requests.

##### `options`

In addition to the options for synchronous decoding, the following options can be specified:

`signal?: AbortSignal`
:   An abort signal to interrupt the stream.

#### Return value

A `StatefulPromise` that resolves to a JavaScript value.

### Example

#### Fetch API

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";

const response = await fetch("http://localhost:8000/rpc", {
  method: "POST",
  headers: {
    Accept: "application/cbor",
    // ... more headers
  },
  // ... more properties
})

const result = await decodeStream(resp.body);
```
