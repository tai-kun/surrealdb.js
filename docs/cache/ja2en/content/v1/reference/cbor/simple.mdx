---
title: Simple
slug: en/v1/reference/cbor/simple
---

import Badge from "~/components/en/Badge.astro";

`Simple` is a simple value identified by an integer value from 0 to 255.

### Import

```ts
import { Simple } from "@tai-kun/surrealdb/cbor";
```

### `.constructor()`

Creates a new `Simple` object.

```ts
new Simple(value: number);
```

#### Arguments

##### `value`

An integer value from 0 to 255.

#### Return Value

When called via `new`, `Simple` returns its instance.

### `.value` <Badge variant="instance" /><Badge variant="property" />

The integer value from 0 to 255 that identifies the `Simple` object.

### Notes

Refer to the following materials:

- [https://datatracker.ietf.org/doc/html/rfc8949#name-simple-values](https://datatracker.ietf.org/doc/html/rfc8949#name-simple-values)
- [https://www.iana.org/assignments/cbor-simple-values/cbor-simple-values.xhtml](https://www.iana.org/assignments/cbor-simple-values/cbor-simple-values.xhtml)
