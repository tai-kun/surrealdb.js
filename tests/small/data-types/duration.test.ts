import { isDuration } from "@tai-kun/surrealdb";
import { decode, encode } from "@tai-kun/surrealdb/cbor";
import { CBOR_TAG_DURATION } from "@tai-kun/surrealdb/encodable-datatypes";
import { Duration } from "@tai-kun/surrealdb/standard-datatypes";
import { describe, expect, test } from "vitest";

const valid = Object.entries({
  0: {
    input: 0,
    toPlainObject: {
      seconds: 0n,
      nanoseconds: 0,
    },
    total: {
      ns: 0n,
      ms: 0,
    },
    json: "0ns",
    surql: "0ns",
    string: "0ns",
    iso8601: "P0S",
  },
  "12345ns": {
    input: "12345ns",
    toPlainObject: {
      seconds: 0n,
      nanoseconds: 12_345,
    },
    total: {
      ns: 12_345n,
      ms: 0.012_345,
    },
    json: "12µs345ns",
    surql: "12µs345ns",
    string: "12µs345ns",
    iso8601: "P0S",
  },
  "012345us": {
    input: "012345us",
    toPlainObject: {
      seconds: 0n,
      nanoseconds: 12_345_000,
    },
    total: {
      ns: 12_345_000n,
      ms: 12.345_000,
    },
    json: "12ms345µs",
    surql: "12ms345µs",
    string: "12ms345µs",
    iso8601: "P0S",
  },
  "012345µs": {
    input: "012345µs",
    toPlainObject: {
      seconds: 0n,
      nanoseconds: 12_345_000,
    },
    total: {
      ns: 12_345_000n,
      ms: 12.345_000,
    },
    json: "12ms345µs",
    surql: "12ms345µs",
    string: "12ms345µs",
    iso8601: "P0S",
  },
  "012345μs": {
    input: "012345μs",
    toPlainObject: {
      seconds: 0n,
      nanoseconds: 12_345_000,
    },
    total: {
      ns: 12_345_000n,
      ms: 12.345_000,
    },
    json: "12ms345µs",
    surql: "12ms345µs",
    string: "12ms345µs",
    iso8601: "P0S",
  },
  "123456ms": {
    input: "123456ms",
    toPlainObject: {
      seconds: 123n,
      nanoseconds: 456_000_000,
    },
    total: {
      ns: 123_456_000_000n,
      ms: 123_456,
    },
    json: "2m3s456ms",
    surql: "2m3s456ms",
    string: "2m3s456ms",
    iso8601: "P2M3S",
  },
  "3601s": {
    input: "3601s",
    toPlainObject: {
      seconds: 3601n,
      nanoseconds: 0,
    },
    total: {
      ns: 3_601_000_000_000n,
      ms: 3_601_000,
    },
    json: "1h1s",
    surql: "1h1s",
    string: "1h1s",
    iso8601: "P1H1S",
  },
  "10000m": {
    input: "10000m",
    toPlainObject: {
      seconds: 600_000n,
      nanoseconds: 0,
    },
    total: {
      ns: 600_000_000_000_000n,
      ms: 600_000_000,
    },
    //  (6 * 24 * 60) + (22 * 60) + (40)
    // = 8_640        +  1_320    +  40
    // = 10_000
    json: "6d22h40m",
    surql: "6d22h40m",
    string: "6d22h40m",
    iso8601: "P6D22H40M",
  },
  "90d": {
    input: "90d",
    toPlainObject: {
      seconds: 7_776_000n, // 90 * 24 * 60 * 60
      nanoseconds: 0,
    },
    total: {
      ns: 7_776_000_000_000_000n,
      ms: 7_776_000_000,
    },
    json: "12w6d",
    surql: "12w6d",
    string: "12w6d",
    iso8601: "P12W6D",
  },
  "100w": {
    input: "100w",
    toPlainObject: {
      seconds: 60_480_000n, // 100 * 7 * 24 * 60 * 60
      nanoseconds: 0,
    },
    total: {
      ns: 60_480_000_000_000_000n,
      ms: 60_480_000_000,
    },
    //   ((1 * 365) + (47 * 7) + 6) / 7
    // = ( 365      +  329     + 6) / 7
    // = ( 365      +  335        ) / 7
    // =  700                       / 7
    // =  100
    json: "1y47w6d",
    surql: "1y47w6d",
    string: "1y47w6d",
    iso8601: "P1Y47W6D",
  },
  "2y": {
    input: "2y",
    toPlainObject: {
      seconds: 63_072_000n,
      nanoseconds: 0,
    },
    total: {
      ns: 63_072_000_000_000_000n,
      ms: 63_072_000_000,
    },
    json: "2y",
    surql: "2y",
    string: "2y",
    iso8601: "P2Y",
  },
  "1ns1us1ms1s1m": {
    input: "1ns1us1ms1s1m",
    toPlainObject: {
      seconds: 61n,
      nanoseconds: 1_001_001,
    },
    total: {
      ns: 61_001_001_001n,
      ms: 61_001.001_001,
    },
    json: "1m1s1ms1µs1ns",
    surql: "1m1s1ms1µs1ns",
    string: "1m1s1ms1µs1ns",
    iso8601: "P1M1S",
  },
  "1ms1m": {
    input: "1ms1m",
    toPlainObject: {
      seconds: 60n,
      nanoseconds: 1_000_000,
    },
    total: {
      ns: 60_001_000_000n,
      ms: 60_001,
    },
    json: "1m1ms",
    surql: "1m1ms",
    string: "1m1ms",
    iso8601: "P1M",
  },
  "1m1s1m1s1m1s": {
    input: "1m1s1m1s1m1s",
    toPlainObject: {
      seconds: 183n,
      nanoseconds: 0,
    },
    total: {
      ns: 183_000_000_000n,
      ms: 183_000,
    },
    json: "3m3s",
    surql: "3m3s",
    string: "3m3s",
    iso8601: "P3M3S",
  },
  "1234.567890": {
    input: 1234.567890,
    toPlainObject: {
      seconds: 1n,
      nanoseconds: 234_567_890,
    },
    total: {
      ns: 1_234_567_890n,
      ms: 1_234.567_890,
    },
    json: "1s234ms567µs890ns",
    surql: "1s234ms567µs890ns",
    string: "1s234ms567µs890ns",
    iso8601: "P1S",
  },
});

for (const [i, t] of valid) {
  describe(String(i), () => {
    test("プロパティー", () => {
      const d = new Duration(t.input);

      expect({ seconds: d.seconds, nanoseconds: d.nanoseconds })
        .toStrictEqual(t.toPlainObject);
    });

    test("コンパクト表現", () => {
      const d = new Duration(t.input);

      expect(d.getCompact()).toStrictEqual([
        t.toPlainObject.seconds,
        t.toPlainObject.nanoseconds,
      ]);
    });

    test("ナノ秒", () => {
      const d = new Duration(t.input);

      expect(d.valueOf()).toBe(t.total.ns);
    });

    test("ミリ秒", () => {
      const d = new Duration(t.input);

      expect(+d).toBe(t.total.ms);
    });

    test("文字列", () => {
      const d = new Duration(t.input);

      expect(`${d}`).toBe(t.string);
      expect(d.toJSON()).toBe(t.string);
      expect(d.toSurql()).toBe(t.string);
    });

    test("ISO 形式の文字列", () => {
      const d = new Duration(t.input);

      expect(d.toISOString()).toBe(t.iso8601);
    });

    test("CBOR でエンコード/デコードできる", () => {
      const input = new Duration(t.input);
      const output = new Duration(t.input);
      const bytes = encode(input);
      const dt = decode(bytes, {
        reviver: {
          tagged(t) {
            switch (t.tag) {
              case CBOR_TAG_DURATION:
                return new Duration(t.value as any);

              default:
                return undefined;
            }
          },
        },
      });

      expect(dt).toStrictEqual(output);
    });

    test(".toJSON()", () => {
      const json = new Duration(t.input).toJSON();

      expect(json).toBe(t.json);
    });

    test(".toSurql()", () => {
      const surql = new Duration(t.input).toSurql();

      expect(surql).toBe(t.surql);
    });

    test(".toPlainObject()", () => {
      const toPlainObject = new Duration(t.input).toPlainObject();

      expect(toPlainObject).toStrictEqual(t.toPlainObject);
    });
  });
}

test("有効な正の秒数で更新", () => {
  const d = new Duration(0);
  d.seconds = 100n;

  expect(d.toString()).toBe("1m40s");
});

test("無効な正の秒数で更新するとエラー", () => {
  const d = new Duration(0);

  expect(() => d.seconds = 2n ** 64n).toThrowErrorMatchingSnapshot();
});

test("負の秒数で更新するとエラー", () => {
  const d = new Duration(0);

  expect(() => d.seconds = -1n).toThrowErrorMatchingSnapshot();
});

test("有効な正のナノ秒数で更新", () => {
  const d = new Duration(0);
  d.nanoseconds = 100;

  expect(d.toString()).toBe("100ns");
});

test("NaN でナノ秒数を更新するとエラー", () => {
  const d = new Duration(0);

  expect(() => d.nanoseconds = NaN).toThrowErrorMatchingSnapshot();
});

test("Infinity でナノ秒数を更新するとエラー", () => {
  const d = new Duration(0);

  expect(() => d.nanoseconds = Infinity).toThrowErrorMatchingSnapshot();
});

test("-Infinity でナノ秒数を更新するとエラー", () => {
  const d = new Duration(0);

  expect(() => d.nanoseconds = -Infinity).toThrowErrorMatchingSnapshot();
});

test("Number.MAX_SAFE_INTEGER + 1 でナノ秒数を更新するとエラー", () => {
  const d = new Duration(0);

  expect(() => d.nanoseconds = Number.MAX_SAFE_INTEGER + 1)
    .toThrowErrorMatchingSnapshot();
});

test(".addYears()", () => {
  const d = new Duration(0);
  d.addYears(1);

  expect(d.toString()).toBe("1y");
});

test(".addWeeks()", () => {
  const d = new Duration(0);
  d.addWeeks(1);

  expect(d.toString()).toBe("1w");
});

test(".addDays()", () => {
  const d = new Duration(0);
  d.addDays(1);

  expect(d.toString()).toBe("1d");
});

test(".addHours()", () => {
  const d = new Duration(0);
  d.addHours(1);

  expect(d.toString()).toBe("1h");
});

test(".addMinutes()", () => {
  const d = new Duration(0);
  d.addMinutes(1);

  expect(d.toString()).toBe("1m");
});

test(".addSeconds()", () => {
  const d = new Duration(0);
  d.addSeconds(1);

  expect(d.toString()).toBe("1s");
});

test(".addMilliseconds()", () => {
  const d = new Duration(0);
  d.addMilliseconds(1);

  expect(d.toString()).toBe("1ms");
});

test(".addMicroseconds()", () => {
  const d = new Duration(0);
  d.addMicroseconds(1);

  expect(d.toString()).toBe("1µs");
});

test(".addNanoseconds()", () => {
  const d = new Duration(0);
  d.addNanoseconds(1);

  expect(d.toString()).toBe("1ns");
});

test(".asYears()", () => {
  const d = new Duration("1y73d");

  expect(d.asYears()).toBe(1.2);
});

test(".asWeeks()", () => {
  const d = new Duration("14d");

  expect(d.asWeeks()).toBe(2);
});

test(".asDays()", () => {
  const d = new Duration("1w");

  expect(d.asDays()).toBe(7);
});

test(".asHours()", () => {
  const d = new Duration("1w3d");

  expect(d.asHours()).toBe(240);
});

test(".asMinutes()", () => {
  const d = new Duration("2m59s1000ms");

  expect(d.asMinutes()).toBe(3);
});

test(".asSeconds()", () => {
  const d = new Duration("500ns");

  expect(d.asSeconds()).toBe(0.000_000_5);
});

test(".asMilliseconds()", () => {
  const d = new Duration("0s");

  expect(d.asMilliseconds()).toBe(0);
});

test(".asMicroseconds()", () => {
  const d = new Duration("1ms100ns");

  expect(d.asMicroseconds()).toBe(1_000.1);
});

test(".asNanoseconds()", () => {
  const d = new Duration("999999999ns");

  expect(d.asNanoseconds()).toBe(999_999_999);
});

describe("Duration.MAX", () => {
  test("毎度異なるインスタンス", () => {
    expect(Duration.MAX).not.toBe(Duration.MAX);
  });

  test("継承先のクラスを使ってインスタンスを作成する", () => {
    class MyDuration extends Duration {}

    expect(MyDuration.MAX).toBeInstanceOf(MyDuration);
  });

  test("秒数は uint64 の最大値", () => {
    expect(Duration.MAX.seconds).toBe(2n ** 64n - 1n);
  });

  test("ナノ秒数は 999,999,999", () => {
    expect(Duration.MAX.nanoseconds).toBe(999_999_999);
  });

  test("年月日等をそれぞれ取得できる", () => {
    expect(Duration.MAX.getYears()).toBe(584942417355);
    expect(Duration.MAX.getWeeks()).toBe(3);
    expect(Duration.MAX.getDays()).toBe(5);
    expect(Duration.MAX.getHours()).toBe(7);
    expect(Duration.MAX.getMinutes()).toBe(0);
    expect(Duration.MAX.getSeconds()).toBe(15);
    expect(Duration.MAX.getMilliseconds()).toBe(999);
    expect(Duration.MAX.getMicroseconds()).toBe(999);
    expect(Duration.MAX.getNanoseconds()).toBe(999);
  });
});

describe("Duration.ZERO", () => {
  test("毎度異なるインスタンス", () => {
    expect(Duration.ZERO).not.toBe(Duration.ZERO);
  });

  test("継承先のクラスを使ってインスタンスを作成する", () => {
    class MyDuration extends Duration {}

    expect(MyDuration.ZERO).toBeInstanceOf(MyDuration);
  });

  test("秒数は 0n", () => {
    expect(Duration.ZERO.seconds).toBe(0n);
  });

  test("ナノ秒数は 0", () => {
    expect(Duration.ZERO.nanoseconds).toBe(0);
  });

  test("年月日等をそれぞれ取得できる", () => {
    expect(Duration.ZERO.getYears()).toBe(0);
    expect(Duration.ZERO.getWeeks()).toBe(0);
    expect(Duration.ZERO.getDays()).toBe(0);
    expect(Duration.ZERO.getHours()).toBe(0);
    expect(Duration.ZERO.getMinutes()).toBe(0);
    expect(Duration.ZERO.getSeconds()).toBe(0);
    expect(Duration.ZERO.getMilliseconds()).toBe(0);
    expect(Duration.ZERO.getMicroseconds()).toBe(0);
    expect(Duration.ZERO.getNanoseconds()).toBe(0);
  });
});

test(".years()", () => {
  expect(Duration.years(1).toString()).toBe("1y");
});

test(".weeks()", () => {
  expect(Duration.weeks(1).toString()).toBe("1w");
});

test(".days()", () => {
  expect(Duration.days(1).toString()).toBe("1d");
});

test(".hours()", () => {
  expect(Duration.hours(1).toString()).toBe("1h");
});

test(".minutes()", () => {
  expect(Duration.minutes(1).toString()).toBe("1m");
});

test(".seconds()", () => {
  expect(Duration.seconds(1).toString()).toBe("1s");
});

test(".milliseconds()", () => {
  expect(Duration.milliseconds(1).toString()).toBe("1ms");
});

test(".microseconds()", () => {
  expect(Duration.microseconds(1).toString()).toBe("1µs");
});

test(".nanoseconds()", () => {
  expect(Duration.nanoseconds(1).toString()).toBe("1ns");
});

// test("年月日等で表現されたミリ秒を取得できる", () => {
//   const d = new Duration(365 * 24 * 60 * 60 * 1_000 + 0.123456);
//   // TODO(tai-kun): expect().toBeCloseTo() でテストして良いか？
// });

test(".clone() で複製できる", () => {
  class MyDuration extends Duration {}

  const d1 = new MyDuration([123, 456]);
  const d2 = d1.clone();

  expect(d2).toBeInstanceOf(MyDuration);
  expect(d1).not.toBe(d2);
  expect(d1.getCompact()).toStrictEqual(d2.getCompact());
});

test("Duration であると判定できる", () => {
  expect(isDuration(new Duration(0))).toBe(true);
  expect(isDuration(0)).toBe(false);
});
