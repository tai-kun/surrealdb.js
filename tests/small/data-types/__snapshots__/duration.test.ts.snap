// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`-Infinity でナノ秒数を更新するとエラー 1`] = `
{
  "actual": -Infinity,
  "integer": true,
  "message": "Value -Infinity is out of range. Expected value to be between 0 and 9007199254740991 as an integer.",
  "name": "NumberRangeError",
  "range": [
    0,
    9007199254740991,
  ],
}
`;

exports[`Infinity でナノ秒数を更新するとエラー 1`] = `
{
  "actual": Infinity,
  "integer": true,
  "message": "Value Infinity is out of range. Expected value to be between 0 and 9007199254740991 as an integer.",
  "name": "NumberRangeError",
  "range": [
    0,
    9007199254740991,
  ],
}
`;

exports[`NaN でナノ秒数を更新するとエラー 1`] = `
{
  "actual": NaN,
  "integer": true,
  "message": "Value NaN is out of range. Expected value to be between 0 and 9007199254740991 as an integer.",
  "name": "NumberRangeError",
  "range": [
    0,
    9007199254740991,
  ],
}
`;

exports[`Number.MAX_SAFE_INTEGER + 1 でナノ秒数を更新するとエラー 1`] = `
{
  "actual": 9007199254740992,
  "integer": true,
  "message": "Value 9007199254740992 is out of range. Expected value to be between 0 and 9007199254740991 as an integer.",
  "name": "NumberRangeError",
  "range": [
    0,
    9007199254740991,
  ],
}
`;

exports[`無効な正の秒数で更新するとエラー 1`] = `
{
  "actual": 18446744073709551616n,
  "integer": true,
  "message": "Value 18446744073709551616 is out of range. Expected value to be between 0 and 18446744073709551615 as an integer.",
  "name": "NumberRangeError",
  "range": [
    0n,
    18446744073709551615n,
  ],
}
`;

exports[`負の秒数で更新するとエラー 1`] = `
{
  "actual": -1n,
  "integer": true,
  "message": "Value -1 is out of range. Expected value to be between 0 and 18446744073709551615 as an integer.",
  "name": "NumberRangeError",
  "range": [
    0n,
    18446744073709551615n,
  ],
}
`;
